include/measurement_kit/nettest.hpp

```diff
@@ -0,0 +1,1922 @@
+    /// you already used this class to start a nettest. In both cases, log
+    /// messages will be routed, so you will know what happened.
+    bool start_tcp_connect(settings::TcpConnectSettings settings, routers::Router *router) noexcept;
+
+    /// Starts a "Telegram" nettest.
+    /// \param settings contains the settings configuring this nettest.
+    /// \param router is the router used to route nettest events. If the
+    /// router is `nullptr`, we'll use an internal dummy router.
+    /// \return true on success, false on failure. There are two reasons why
+    /// this method could fail. Either the provided settings are not JSON
+    /// serializable because there was a string that was not valid UTF-8, or
+    /// you already used this class to start a nettest. In both cases, log
+    /// messages will be routed, so you will know what happened.
+    bool start_telegram(settings::TelegramSettings settings, routers::Router *router) noexcept;
+
+    /// Starts a "WebConnectivity" nettest.
+    /// \param settings contains the settings configuring this nettest.
+    /// \param router is the router used to route nettest events. If the
+    /// router is `nullptr`, we'll use an internal dummy router.
+    /// \return true on success, false on failure. There are two reasons why
+    /// this method could fail. Either the provided settings are not JSON
+    /// serializable because there was a string that was not valid UTF-8, or
+    /// you already used this class to start a nettest. In both cases, log
+    /// messages will be routed, so you will know what happened.
+    bool start_web_connectivity(settings::WebConnectivitySettings settings, routers::Router *router) noexcept;
+
+    /// Starts a "Whatsapp" nettest.
+    /// \param settings contains the settings configuring this nettest.
+    /// \param router is the router used to route nettest events. If the
+    /// router is `nullptr`, we'll use an internal dummy router.
+    /// \return true on success, false on failure. There are two reasons why
+    /// this method could fail. Either the provided settings are not JSON
+    /// serializable because there was a string that was not valid UTF-8, or
+    /// you already used this class to start a nettest. In both cases, log
+    /// messages will be routed, so you will know what happened.
+    bool start_whatsapp(settings::WhatsappSettings settings, routers::Router *router) noexcept;
+
+    // Event processing methods
+    // ````````````````````````
+
+    /// Blocks until the next event occurs, then routes it.
+    /// \param router Router to handle events. If the router is null, we'll
+    /// use an internal dummy router.
+    void route_next_event(routers::Router *router) noexcept;
+
+    /// Returns true if nettest is done, false otherwise.
+    bool is_done() noexcept;
+
+    /// Interrupts the currently running nettest.
+    void interrupt() noexcept;
+
+  private:
+    // Private methods
+    // ```````````````
```

Comment:
The public implementation should be such that we could possibly even migrate away from all the JSON dance below and there is not implementation coupling (I think) with such details. This is good because for most of our use cases I noticed that we can actually use SWIG on top of this API rather than using JSON. The advantage is that using this API there is much more certainty with respect to types.

(I am not suggesting that we do that, I am just suggesting that exposing more JSON details'd be bad.)